### PR TITLE
jit: Turn various type hint bugs into compile-time errors

### PR DESCRIPTION
**Pull request by jhogberg**
We've had a few too many bugs where we used `BEAM_TYPE_INTEGER | BEAM_TYPE_MASK_BOXED` instead of  `BEAM_TYPE_INTEGER | BEAM_TYPE_ALWAYS_BOXED`.

This PR turns that kind of error, and a few other misuses, into compile-time errors.

### CODE REVIEW

File: erts/emulator/beam/jit/beam_jit_common.hpp [hunk 1]
```diff
@@ -37,8 +37,100 @@ extern "C"
...
+             static_cast<integral>(T)) == BEAM_TYPE_NONE;
+    static constexpr bool no_boxed_overlap =
+            no_overlap || (super::value() != BeamTypeId::MaybeBoxed &&
+                           T != BeamTypeId::MaybeBoxed);
+    static constexpr bool no_immed_overlap =
+            no_overlap || (super::value() != BeamTypeId::MaybeImmediate &&
+                           T != BeamTypeId::MaybeImmediate);
+
+    static_assert(no_boxed_overlap,
+                  "types may not overlap, did you mean to use "
```
Comment (bjorng) [inline erts/emulator/beam/jit/beam_jit_common.hpp:114]:
```suggestion
                  "types must not overlap, did you mean to use "
```

File: erts/emulator/beam/jit/beam_jit_common.hpp [hunk 2]
```diff
@@ -37,8 +37,100 @@ extern "C"
...
+                  "BeamTypeId::AlwaysBoxed here?");
+    static_assert(no_immed_overlap,
+                  "types may not overlap, did you mean to use "
```
Comment (bjorng) [inline erts/emulator/beam/jit/beam_jit_common.hpp:117]:
```suggestion
                  "types must not overlap, did you mean to use "
```

File: erts/emulator/beam/jit/beam_jit_common.hpp [hunk 3]
```diff
@@ -37,8 +37,100 @@ extern "C"
...
+                  "BeamTypeId::AlwaysImmediate here?");
+    static_assert(no_overlap || no_boxed_overlap || no_immed_overlap,
+                  "types may not overlap");
```
Comment (bjorng) [inline erts/emulator/beam/jit/beam_jit_common.hpp:120]:
```suggestion
                  "types must not overlap");
```

File: erts/emulator/beam/jit/x86/instr_common.cpp [hunk 4]
```diff
@@ -1961,14 +1964,14 @@ void BeamModuleAssembler::emit_is_in_range(ArgLabel const &Small,
...
+    } else if (always_one_of<BeamTypeId::Integer, BeamTypeId::AlwaysBoxed>(
+                       Src)) {
         /* The only possible kind of immediate is a small and all
          * other values are boxed, so we can test for smalls by
          * testing boxed. */
         comment("simplified small test since all other types are boxed");
         ERTS_CT_ASSERT(_TAG_PRIMARY_MASK - TAG_PRIMARY_BOXED == (1 << 0));
-        if (Small == Large &&
-            always_one_of(Src, BEAM_TYPE_MASK_BOXED - BEAM_TYPE_FLOAT)) {
+        if (Small == Large && never_one_of<BeamTypeId::Float>(Src)) {
```
Comment (bjorng) [inline erts/emulator/beam/jit/x86/instr_common.cpp:1974]: You can simplify the surrounding code by using `emit_is_not_boxed`.

**Review by bjorng — APPROVED** [review]
Looks good to me. I like how much clearer the type tests now read.

File: erts/emulator/beam/jit/arm/instr_common.cpp [hunk 5]
```diff
@@ -1926,14 +1929,14 @@ void BeamModuleAssembler::emit_is_in_range(ArgLabel const &Small,
...
+    } else if (always_one_of<BeamTypeId::Integer, BeamTypeId::AlwaysBoxed>(
+                       Src)) {
         /* The only possible kind of immediate is a small and all
          * other values are boxed, so we can test for smalls by
          * testing boxed. */
         comment("simplified small test since all other types are boxed");
         ERTS_CT_ASSERT(_TAG_PRIMARY_MASK - TAG_PRIMARY_BOXED == (1 << 0));
-        if (Small == Large &&
-            always_one_of(Src, BEAM_TYPE_MASK_BOXED - BEAM_TYPE_FLOAT)) {
+        if (Small == Large && never_one_of<BeamTypeId::Float>(Src)) {
```
[the remark this answers is not in the document]
Comment (jhogberg) [reply erts/emulator/beam/jit/arm/instr_common.cpp:1939]: Thanks, is it alright if I do it in a different PR? I want this merged as soon as possible to remove it as a blocker for the merge of #6733 and testing of #6731.

[the remark this answers is not in the document]
Comment (bjorng) [reply erts/emulator/beam/jit/arm/instr_common.cpp:1939]: Yes, a different PR is alright.

**Pull request merged**